spynnaker/pyNN/models/neuron/synaptic_matrix_app.py

```diff
@@ -0,0 +1,719 @@
+            The space available for "direct" or "single" synapses
+        :param ~pacman.model.graphs.common.Slice post_vertex_slice:
+            The slice of the post-vertex the matrix is for
+        :param int synaptic_matrix_region:
+            The region where synaptic matrices are stored
+        :param int direct_matrix_region:
+            The region where "direct" or "single" synapses are stored
+        """
+        self.__synapse_io = synapse_io
+        self.__poptable = poptable
+        self.__synapse_info = synapse_info
+        self.__app_edge = app_edge
+        self.__n_synapse_types = n_synapse_types
+        self.__all_single_syn_sz = all_single_syn_sz
+        self.__post_vertex_slice = post_vertex_slice
+        self.__synaptic_matrix_region = synaptic_matrix_region
+        self.__direct_matrix_region = direct_matrix_region
+
+        # Map of machine_edge to .SynapticMatrix
+        self.__matrices = dict()
+
+        # Calculate the max row info for this edge
+        self.__max_row_info = self.__synapse_io.get_max_row_info(
+            synapse_info, self.__post_vertex_slice,
+            app_edge.n_delay_stages, self.__poptable,
+            globals_variables.get_simulator().machine_time_step, app_edge)
+
+        # These are set directly later
+        self.__all_syn_block_sz = None
+        self.__app_key_info = None
+        self.__delay_app_key_info = None
+        self.__routing_info = None
+        self.__weight_scales = None
+        self.__m_edges = None
+        self.__use_app_keys = None
+
+        self.__matrix_size = (
+            self.__app_edge.pre_vertex.n_atoms *
+            self.__max_row_info.undelayed_max_bytes)
+        self.__delay_matrix_size = (
+            self.__app_edge.pre_vertex.n_atoms *
+            self.__app_edge.n_delay_stages *
+            self.__max_row_info.delayed_max_bytes)
+        vertex = self.__app_edge.pre_vertex
+        self.__n_subatoms = int(min(
+            vertex.get_max_atoms_per_core(), vertex.n_atoms))
+        self.__n_subedges = int(math.ceil(vertex.n_atoms / self.__n_subatoms))
+
+        # These are computed during synaptic generation
+        self.__syn_mat_offset = None
+        self.__delay_syn_mat_offset = None
+        self.__index = None
+        self.__delay_index = None
+
+        # These are stored when blocks are read
+        self.__received_block = None
+        self.__delay_received_block = None
+
+    def __get_matrix(self, machine_edge):
+        """ Get or create a matrix object
+
+        :param ~pacman.model.graph.machine.machine_edge.MachineEdge
+            machine_edge: The machine edge to get the matrix for
+        :rtype: SynapticMatrix
+        """
+        if machine_edge in self.__matrices:
+            return self.__matrices[machine_edge]
+
+        r_info = self.__routing_info.get_routing_info_for_edge(machine_edge)
+        delayed_r_info = None
+        delayed_edge = machine_edge.delay_edge
```

Comment:
delayed app or delayed mac edge? name ambigious

Reply:
There's a better way to fix that... done